lgc/tool/lgc/lgc.cpp

```diff
@@ -0,0 +1,259 @@
+ * @file  lgc.cpp
+ * @brief LLPC source file: contains implementation of LGC standalone tool.
+ ***********************************************************************************************************************
+ */
+
+#include "lgc/LgcContext.h"
+#include "lgc/Pipeline.h"
+#include "llvm/AsmParser/Parser.h"
+#include "llvm/BinaryFormat/ELF.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/SourceMgr.h"
+
+using namespace lgc;
+using namespace llvm;
+
+// Input sources
+static cl::list<std::string> InFiles(cl::Positional, cl::OneOrMore, cl::ValueRequired,
+                                     cl::desc("Input file(s) (\"-\" for stdin)"));
+
+// -extract: extract a single module from a multi-module input file
+static cl::opt<unsigned> Extract("extract", cl::desc("Extract single module from multi-module input file"), cl::init(0),
+                                 cl::value_desc("1-based index of module to extract"));
+
+// -o: output filename
+static cl::opt<std::string> OutFileName("o", cl::desc("Output filename"),
+                                        cl::value_desc("filename (\"-\" for stdout)"));
+
+// -pal-abi-version: PAL pipeline ABI version to compile for (default is latest known)
+static cl::opt<unsigned> PalAbiVersion("pal-abi-version", cl::init(0xFFFFFFFF),
+                                       cl::desc("PAL pipeline version to compile for (default latest known)"));
+
+// =====================================================================================================================
+// Checks whether the input data is actually a ELF binary
+//
+// @param data : Input data to check
+static bool isElfBinary(StringRef data) {
+  bool isElfBin = false;
+  if (data.size() >= sizeof(ELF::Elf64_Ehdr)) {
+    auto pHeader = reinterpret_cast<const ELF::Elf64_Ehdr *>(data.data());
+    isElfBin = pHeader->checkMagic();
+  }
+  return isElfBin;
+}
+
+// =====================================================================================================================
+// Checks whether the output data is actually ISA assembler text
+//
+// @param data : Input data to check
+static bool isIsaText(StringRef data) {
+  // This is called by the lgc standalone tool to help distinguish between its three output types of ELF binary,
+  // LLVM IR assembler and ISA assembler. Here we use the fact that ISA assembler is the only one that starts
+  // with a tab character.
+  return data.startswith("\t");
+}
+// =====================================================================================================================
+// Main code of LGC standalone tool
+//
+// @param argc : Count of command-line arguments
+// @param argv : Command-line arguments
+int main(int argc, char **argv) {
+  const char *progName = sys::path::filename(argv[0]).data();
```

Comment:
nit: const char *const progName

Reply:
We don't generally have a style of consting local variables (although a couple of places written by Neil do it), and I don't think we want to start doing it.